models/classes/CrudResultsService.php

```diff
@@ -55,47 +59,52 @@ public function get($uri, $groupBy = self::GROUP_BY_DELIVERY)
     {
         $deliveryExecution = ServiceProxy::singleton()->getDeliveryExecution($uri);
         $delivery = $deliveryExecution->getDelivery();
-    
+
         $resultService = $this->getServiceLocator()->get(ResultServerService::SERVICE_ID);
         $implementation = $resultService->getResultStorage($delivery->getUri());
         return $this->format($implementation, $uri);
     }
-        
-    public function format(\taoResultServer_models_classes_ReadableResultStorage $resultStorage, $resultIdentifier, $groupBy = self::GROUP_BY_DELIVERY)
+
+    /**
+     * @param taoResultServer_models_classes_ReadableResultStorage $resultStorage
+     * @param $resultIdentifier
+     * @param int $groupBy
+     * @param bool $lastResult
+     * @return array
+     * @throws common_exception_Error
+     */
+    public function format(taoResultServer_models_classes_ReadableResultStorage $resultStorage, $resultIdentifier, $groupBy = self::GROUP_BY_DELIVERY, $lastResult = false)
     {
         $returnData = [];
-        
+
         if ($groupBy === self::GROUP_BY_DELIVERY || $groupBy === self::GROUP_BY_ITEM) {
             $calls = $resultStorage->getRelatedItemCallIds($resultIdentifier);
         } else {
             $calls = $resultStorage->getRelatedTestCallIds($resultIdentifier);
+            $lastResult = false;
         }
 
         foreach ($calls as $callId) {
             $results = $resultStorage->getVariables($callId);
-            $resource = [];
+            $lastData = [];
             foreach ($results as $result) {
                 $result = array_pop($result);
                 if (isset($result->variable)) {
-                    $resource['value'] = $result->variable->getValue();
-                    $resource['identifier'] = $result->variable->getIdentifier();
-                    if ($result->variable instanceof \taoResultServer_models_classes_ResponseVariable) {
-                        $type = "http://www.tao.lu/Ontologies/TAOResult.rdf#ResponseVariable";
-                    } else {
-                        $type = "http://www.tao.lu/Ontologies/TAOResult.rdf#OutcomeVariable";
+                    $resource = $this->formResource($result->variable);
+                    if (!$lastResult) {
+                        $lastData[] = $resource;
+                    } elseif ((empty($lastData[$resource['identifier']])
+                        || preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $resource['epoch'])
+                        > preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $lastData[$resource['identifier']]['epoch']))) {
+                        $lastData[$resource['identifier']] = $resource;
```

Comment:
I would suggest to make this more clear.
if I get correctly we are validating the current resource epoch whether is bigger that the last last one.
What about making extract those on variables and make more clear
Like

```
$currentAttemptTimestampt = preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $resource['epoch']);
$lastReccoredAttemptTimestampt =   preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $lastData[$resource['identifier']]['epoch']);

```
Also I would prefer to have the condition here simpler just
If () else
I would extract the further validation on whether is a last attempt or not on a separate method